src/Features/CSharp/Portable/RemoveUnreachableCode/CSharpRemoveUnreachableCodeDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,161 @@
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeBlock, ImmutableArray.Create(SyntaxKind.Block));
+
+        private void AnalyzeBlock(SyntaxNodeAnalysisContext context)
+        {
+            var block = context.Node;
+            if (!IsOutermostBlock(block))
+            {
+                // Don't bother processing inner blocks.  We'll have already checked them when
+                // we processed the outer block
+                return;
+            }
+
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+
+            // There is no good existing API to check if a statement is unreachable in an efficient
+            // manner.  While there is SemanticModel.AnalyzeControlFlow, it can only operate on a 
+            // statement at a time, and it will reanalyze and allocate on each call.  
+            //
+            // To avoid this, we simply ask the semantic model for all the diagnostics for this
+            // block and we look for any reported "unreachable code detected" diagnostics.
+            //
+            // This is actually quite fast to do because the compiler does not actually need to
+            // recompile things to determine the diagnostics.  It will have already stashed the
+            // binding diagnostics directly on the SourceMethodSymbol containing this block, and
+            // so it can retrieve the diagnostics at practically no cost.
+            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
+            var diagnostics = semanticModel.GetDiagnostics(block.Span, cancellationToken);
+            foreach (var diagnostic in diagnostics)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (diagnostic.Id == CS0162)
+                {
+                    ProcessUnreachableDiagnostic(context, root, diagnostic.Location.SourceSpan);
+                }
+            }
+        }
+
+        private bool IsOutermostBlock(SyntaxNode block)
+        {
+            // Avoid linq/allocations as we will be calling this on every block in the solution.
+            for (var current = block.Parent; current != null; current = current.Parent)
+            {
+                if (current.Kind() == SyntaxKind.Block)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void ProcessUnreachableDiagnostic(
+            SyntaxNodeAnalysisContext context, SyntaxNode root, TextSpan sourceSpan)
+        {
+            var node = root.FindNode(sourceSpan);
+
+            // Note: this approach works as the language only supports the concept of 
```

Comment:
i really don't understand the distinctions your'e drawing here.  Nor do i really think it's necessary to update the language in that manner.  The comment is fairly clear.  This code is designed to work as per the known design of C# that we only have unreachable statements.  If that changes in the future we'll need to revise this.  I dont' think we really need to say more than that.